docs/eeps/draft-class-loading.md

```diff
@@ -0,0 +1,524 @@
+   289 Wed Dec 06 20:29:12 UTC 2017 io/netty/util/ReferenceCounted.class
+  1335 Wed Dec 06 20:29:12 UTC 2017 io/netty/util/Recycler$2.class
+   248 Wed Dec 06 20:29:12 UTC 2017 io/netty/util/ResourceLeak.class
+  5864 Wed Dec 06 20:29:12 UTC 2017 io/netty/util/Version.class
+...
+```
+
+As you can see, not only are Embulk's own classes and resources there, but also Guice, Guava, Jackson, Logback, SLF4J, JRuby, SnakeYAML, MessagePack, ICU4J, and Netty. They were loaded by the top-level class loader and caused a lot of dependency conflicts with plugins.
+
+Built-in plugins were also extracted into the executable JAR files, such as the CSV Parser (`CsvParserPlugin`) and the Local File Input (`LocalFileInputPlugin`) along with the core dependency libraries. They caused the same issues.
+
+[EEP-7](./eep-0007.md) described the problems by the dependency conflicts and the criteria for removing, hiding, or keeping each dependency library. This EEP discusses how to hide the "hidden" dependencies and built-in plugins.
+
+`parent_first_packages` and `parent_first_resources`
+-----------------------------------------------------
+
+Embulk has a mechanism to resolve dependency conflicts called `parent_first_packages`, but it does not actually work properly.
+
+This `parent_first_packages` is a mechanism to suppress some classes from being loaded by plugin class loaders. It is a list of Java package names. When a plugin class loader is asked to load a class listed in `parent_first_packages`, the plugin class loader first looks for the class from the top-level class loader. Otherwise, it loads the class from the plugin class loader.
+
+This has two problems. First, it does not solve the problem of mixing two versions of Jackson add-on modules, which is discussed in [EEP-7](./eep-0007.md). A plugin may want to use `jackson-dataformat-cbor:2.13.5` while the Embulk core still uses `jackson-core:2.6.7`.
+
+Second, this mechanism applies to each class, not each dependency library artifact JAR file. Therefore, a mix-up can occur within a single library. For example, recent Guava 30.0+ has a new class `com.google.common.util.concurrent.ClosingFuture`, while Guava 18.0 used in the Embulk core (up to v0.10.38) did not. If a plugin includes Guava 30.0+, `ClosingFuture` is loaded from that Guava 30.0+. `ClosingFuture` uses another Guava class, `com.google.common.collect.ImmutableList`. `ImmutableList` is loaded from Guava 18.0 in Embulk v0.10.38 or earlier because it existed in Guava 18.0 and `com.google.common` was included in `parent_first_packages` of Embulk v0.10.38 or earlier. When `ClousingFuture` called a new method in `ImmutableList` that was added after Guava 18.0, `NoSuchMethodError` would be thrown.
+
+Because of these two problems, a different mechanism for isolating library artifact JAR files other than `parent_first_packages` was needed.
+
+The `parent_first_resources` file is a similar mechanism for non-class resources, which has the same problems.
+
```

Comment:
`embulk-deps` replaced `parent_first_packages` and `parent_first_resources` mechanisms?
Or It still uses in v0.11.x?

Reply:
Added the following sentense.

> The `parent_first_packages` and `parent_first_resources` mechanisms still remain in Embulk as of v0.11, but they will go away once they are no longer needed.